### ISSUE TITLE
Feature Request: IPv6 Support

### ISSUE THREAD

**Issue by ioannidesalex** [hestiacp/hestiacp#429]
IPv6 Support.

**Comment by JOduMonT** [comment]
I understand IPv6 is on the radar of HestiaCP but
for now if any of us have an IPv6 on his server it's means every services are open to the internet

until this is fixed by having a IPv6 firewall, I suggest to bind mysql to an IP so it is not expose to IPv6

such as I scanned my server with http://www.ipv6scanner.com/cgi-bin/main.py
than MySQL was involuntary OPEN
while simply by adding `bind-address=127.0.0.1` inside my `/etc/mysql/my.cnf`
remark 127.0.0.1 could be replace by the IPv4 of the server or even `bind-address=0.0.0.0` which exclude IPv6.

**Comment by ScIT-Raphael** [comment]
Thanks for pointing us to the potential security issue @JOduMonT, I'll check how we can handle this.

**Comment by JOduMonT** [comment]
> I'll check how we can handle this.

if you want I could do a PR with the `bind-address=0.0.0.0`

**Comment by JOduMonT** [comment]
Since no port are binded SSH and Mail are accessible via IPv6

**NOTE:** for SSH it could be also close by only accepting IPv4 with this parameter: `AddressFamily ipv4`

### On the other hand if Hestia want to move forward instead of backward :)
@ScIT-Raphael is Hestia have a way to detect IPv6 ?
such as %ip% for IPv4 ?
if yes we just need to add 

since I just have one IPv4 I simply add those line to my nginx config
`    listen      [::]:%web_port%;`
`    listen      [::]:%web_ssl_port% ssl http2;`

#### so now my Hestia is IPv6 almost ready
### what is missing ?
- fail2ban
- firewall

**Comment by ScIT-Raphael** [comment]
We discussed this internaly and would like to prefer the way to add firewall support for ipv6, if we would change the bind address in mysql configuration, we could break some configurations of our clients. I'll work on the ipv6 implementation anyway with priority now.

**Comment by JOduMonT** [comment]
> add firewall support for ipv6

sound great, do you when it will be implemented ?

**Comment by ScIT-Raphael** [comment]
No, priority is now to ship the new release (1.1.0), after that we will get some free ressources.

**Comment by atik7** [comment]
Please support IPv6!

**Comment by MrTheSoulz** [comment]
I would also love to see ipv6 supported. This is a somewhat big deal for me, its something i missed on VestaCP and i was really hopping i would get it by changing to HestiaCP.

**Comment by bisubus** [comment]
Regarding #691, the CP has reasonable default configs for the most part and I was surprised to see that MySQL is exposed, which is a security concern. I'd expect `bind-address=127.0.0.1` to be a default but there's even no commented line. Since no fine-tuning is provided during installation, this could be mentioned a couple of times at least.

**Comment by ioannidesalex** [comment]
How is MySQL exposed?

**Comment by jaapmarcus** [comment]
Open port with IPv6

```
tcp6       0      0 :::3306                 :::*                    LISTEN      980/mariadbd        
tcp6       0      0 :::110                  :::*                    LISTEN      740/dovecot         
tcp6       0      0 ::1:783                 :::*                    LISTEN      981/perl            
tcp6       0      0 :::143                  :::*                    LISTEN      740/dovecot         
tcp6       0      0 :::53                   :::*                    LISTEN      751/named           
tcp6       0      0 :::22                   :::*                    LISTEN      7792/sshd           
tcp6       0      0 ::1:953                 :::*                    LISTEN      751/named           
tcp6       0      0 :::993                  :::*                    LISTEN      740/dovecot         
tcp6       0      0 :::995                  :::*                    LISTEN      740/dovecot 
```

The firewalll is fully open for IPv6

**Comment by ioannidesalex** [comment]
I see. I thought @bisubus was referring to IPv4.

I am not using IPv6 so not really aware of this.

**Comment by ScIT-Raphael** [comment]
To put the problem into perspective: By default, port 3306 is accessible from the outside of Vesta. At Hestia we decided to block the port via firewall to prevent potential brute force attacks.

The problem has also already been addressed in our forum: https://forum.hestiacp.com/t/hestia-1-3-0-rc-ports-open-for-ipv6/1904

Since the service has an authentication layer, we do not see a critical problem here. We will now focus on the release of the new version 1.3.0, tackle IPv6 and prioritize the firewall implementation.

**Comment by bisubus** [comment]
@ioannidesalex I referred to IPv4 indeed. To make it clear, in 1.2 installation with iptables+fail2ban, 3306 was opened by Maria and it wasn't closed by the firewall, there was no DROP rule by default in iptables. I was able to connect from remote machine to DB before closing it manually. I hijacked this issue because the problem with 3306 has already been addressed here.

**Comment by ioannidesalex** [comment]
@bisubus As far as I know, default policy for our IPtables configuration is DROP by default. Which means that if there is no rule in Firewall it drops the inbound connection. You don't have the line :INPUT DROP [0:0] in your /etc/iptables.rules file?

By the way, according to your previous post/suggestion, if we add bind-address=0.0.0.0 it means that MariaDB listens to all IPs anyway. The safest way is to use bind-address=127.0.0.1 in order to avoid any remote connections. However, blocking port 3306 should be enough.

On another matter, having phpMyAdmin is somehow exposing database anyway :)

**Comment by Lupul** [comment]
I can confirm the default policy for the INPUT chain is DROP
> $ iptables -n -L INPUT -t filter
**Chain INPUT (policy DROP)**
target     prot opt source               destination         
fail2ban-HESTIA  tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:8083
fail2ban-SSH  tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:39178
fail2ban-MAIL  tcp  --  0.0.0.0/0            0.0.0.0/0            multiport dports 25,465,587,110,995,143,993
fail2ban-RECIDIVE  tcp  --  0.0.0.0/0            0.0.0.0/0            multiport dports 1:65535
ACCEPT     all  --  0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED
...
```

**Comment by bisubus** [comment]
Yes, that was a typo, `bind-address=127.0.0.1` is what I meant and actually did.
It appears that the primary cause is that /etc/iptables.rules exists and DROP is there but it's not restored after reboot. In my case it's `Chain INPUT (policy ACCEPT)`. For some reason iptables-restore isn't called despite it was written to configs by Hestia installer. The problem was on my side, iptables-restore was supposed to be performed on networkd-dispatcher start that didn't happen because it depends on systemd-networkd, which was disabled for some reason. Thanks for the assistance on that and sorry for the disturbance.

**Comment by ipkpjersi** [comment]
Do we know if IPv6 support is planned for 2021, or maybe 2022?

**Comment by jaydio** [comment]
Any update on this? I'd love to see this on the roadmap for early next year. More than happy to throw a bounty your way if that helps to prioritize this?

**Comment by ManuelReschke** [comment]
How can we help? It would be awesome to have IPv6 support.

**Comment by ipkpjersi** [comment]
I'd also be willing to throw a (pretty sizable) bounty to help prioritize this for whoever wants to work on it, it'd be super cool to see IPv6 support.

**Comment by jult** [comment]
You can always install CSF/LFD as your firewall, it handles ipv6 just fine. That's what I've been doing for years already. Don't use hestia for firewall config, LFD also replaces fail2ban if you config it right. You can have CSF web-UI access through a protected nginx proxy if so desired.

**Comment by ipkpjersi** [comment]
I think the idea of hestia is that people don't want to have to replace core components of it, and then beyond that even have to configure them more than that.

**Comment by jult** [comment]
Not entirely true, you have the option to not install a firewall or dns control using hestiaCP, it's just what your preference may be. I never liked a control panel for firewalling tasks, it's way too risky/mundane if that could be handled by a control panel. Furthermore, if you already use CSF/LFD elsewhere, it's easy to cluster their config. I don't see myself doing that with databases or mail or user or web management.

**Comment by jult** [comment]
Or, do as I do, install and use CSF/LFD instead of other firewalls/fail2ban. It already rejects all ipv6 inwards, if you haven't opened ports up manually in csf; https://configserver.com/configserver-security-and-firewall/
CSF is - in my experience - by far the best linux firewalling around today. Its documentation is superb, it has a nice UI and its alerting system is easy to config.

**Comment by Cr0iX** [comment]
Hope to see ipv6 support soon.

**Comment by bugsse** [comment]
What's the status here? Ipv6 support has become a serious deal breaker and I have actually started moving to other panels due to the lack of ipv6 in Hestia.  As much as I love Hestia most of my traffic is now over ipv6, it is 2023 after all.

What would it cost to have someone make this happen?

**Comment by jult** [comment]
In the whole of France the mobile network operators/ISPs are all native IPv6 now. They still allow IPv4, but default to IPv6 everywhere. It took me some time to find out why some of my users could not access their email and/or websites while in France. Pretty drastic, but it tells you the urgency of ipv6 support.
I moved all my users to ISPconfig, after [my communication with one of the hestia devs](https://github.com/hestiacp/hestiacp/pull/3882#issuecomment-1692070244) about it.

**Comment by cantalupo555** [comment]
Any IPv6 news?

**Comment by dinandmentink** [comment]
Question. Hope someone can help me out while we wait for IPV6 to be available by default in hestia. 

I am trying to add a template which enables ipv6, but for the life of me can´t  get it to work. Can someone spot my mistake? This is what I did.

In `/usr/local/hestia/data/templates/web/nginx`:

```
cp default.tpl ipv6.tpl
cp default.stpl ipv6.stpl
```

Add ipv6 listen directive:

```
# ipv6.tpl
server {
        listen      [::]:%proxy_port%;
        listen      %ip%:%proxy_port%;
...
```

```
# ipv6.stpl
server {
        listen      [::]:%proxy_ssl_port% ssl http2;
        listen      %ip%:%proxy_ssl_port% ssl;
...
```

Then set the domain to ipv6 and rebuild. For some reason accessing the site over ipv6 still doesn't work (tested with https://ready.chair6.net/ and https://internet.nl).  Am I missing something obvious?

I have a firewall outside of hestia, so mysql ports and such are not a concern for me.

**Comment by linkp** [comment]
The [HestiaCP Forum](https://forum.hestiacp.com/) is a more appropriate venue for your proposed discussion, @dinandmentink.

**Comment by jaydio** [comment]
Any news on this front?

**Comment by ScIT-Raphael** [comment]
There is some progress but we are all hardly limited with our ressources due to private life (family, jobs and so on). You can find the branch here: https://github.com/hestiacp/hestiacp/tree/ipv6

**Comment by Plansol** [comment]
Hi, any updates on this matter? 
Thanks

**Comment by PacaPop** [comment]
Is this still under development? It's becoming a critical issue since some apple devices are already marking websites without IPV6 access as unsafe... and inaccessible.

**Comment by Dark-Sheepa** [comment]
I also struggled with this today, trying to get it working at least provisionally on my server. Even though I got it working provisionally, outside of CP. I can't say I'm satisfied with it, because my provisional solution has plenty of bugs. But I had to do it because people who no longer have access to IPv4 addresses at all _(which some providers no longer have)_ were already visiting my website. From what I've seen, there is already a "pull request," but I have no idea what stage it's at right now. [Pull requests 48](https://github.com/hestiacp/hestiacp/pull/4996)

**Comment by PacaPop** [comment]
@Dark-Sheepa I solved this by setting up Cloudflare on all my websites inside hestia. Cloudflare somehow gives all your websites an IPv4 and IPv6.

**Comment by Dark-Sheepa** [comment]
That sounds like a pretty interesting alternative until we finally get full IPv6 support here. Anything (even if it's only partial emulation) is better than not having IPv6 at all. I understand that in 2019 it wasn't such a big topic, so it wasn't that urgent, but six years later it's quite a big issue. A few years ago, I worked for an internet provider for a while, and even then we had to fight tooth and nail for IPv4 blocks just to get any at all. Fortunately, most things now have full IPv6 support, so it's not as much of an issue as it used to be. But hopefully, things are on the right track.

**Comment by PacaPop** [comment]
Yeap. Cloudflare has been a life saver for me. I couldnt recommend it enough for security reasons... And its free. i even created some custom fail2ban rules that connect with cloudflare api and ban some annoying bots.

**Comment by Dark-Sheepa** [comment]
Well, I found out that Cloudflare is not the right solution for me, and I'm back to where I started. My problem is that Cloudflare proxy limits the "Maximum Upload Size" to 100 MB in the free plan, and even in the highest and most expensive plan, they still ONLY offer 500 MB, which is nowhere near enough for my website. As I have already mentioned in another thread, I am a game developer, and my colleagues upload updates to our game to the website, which our Patreons download. Each update requires over 1 GB to upload via the web interface, because my colleagues definitely cannot access the server, as I have completely disabled FTP, SFTP, and others to minimize threats, and the only way is to upload via the web interface.

It's a shame that the authors themselves are unable to write here how much they would need in donations to make the update to IPv6 so necessary. If they are not interested in this, I will unfortunately have to reinstall the server and install ISPConfig. There's nothing I can do about it, but I gave it a chance. Maybe sometime in the future it will wake up from hibernation.

**Comment by linkp** [comment]
The only way the 100MB upload limit of the Cloudflare free plan will have any effect is if your application does not use the long-established standard practice of chunked uploads. If you fix your application's upload architecture to use chunks that are sized under the limit, it will pose no negative impact.

**Comment by Dark-Sheepa** [comment]
I know, but I'm not a programmer to do that. I'm a network engineer and 3D graphic designer/animator. The system we use is closed source, and the person who custom-built it for us said that it wouldn't be worth the time. Therefore, the quickest solution was to switch to ISPConfig, which already fully supports IPv6. I just installed it, so my problem is solved.

I was willing to send the authors a very generous donation in the thousands, but if they don't communicate... What can I say? 🤷🏻

**Comment by PetarPetrushev** [comment]
Any updates on ipv6 support?

**Comment by asmcc** [comment]
Yes, I'm back here after a break of 2-3 years and I'm continuing to work on the IPV6 topic. Meanwhile @coriaweb continued the work ([see PR4996](https://github.com/hestiacp/hestiacp/pull/4996)) during my interruption, but it could not be implemented directly in the main branch due to complexity and some problems with tests. More details see in the above cited PR. Now, this PR is outdated and it is almost impossible to merge it directly in the main branch. Therefore I reviewed all the changes from the above cited PR of @coriaweb and from his forked repository and merge it now step-by-step with my old IPV6 specific stuff. That is not so simple and requires a lot of time. In parallel I actualised the IPV6 branch to the actual main and try to hold it in sync with the main branch now. And also this "syncing" is not so simple, if you have changes in the main branch in parallel and have to resolve the conflicts manually.
Rudimentary functions work now. You can install it (but not update yet), you can add/delete IP, you can add/delete/change web, user, mail, dns. And yes, I resolved the above discussed problem with IPV6 specific templates for NGINX, APACHE, etc. But a lot of work is not completed yet. @coriaweb worked for example on IPV6 specific firewall. This part was not implemented in my old version. And that is my next topic, which I want to implement step-by-step in the next weeks.

<img width="1045" height="544" alt="Image" src="https://github.com/user-attachments/assets/e260ae6d-88c7-4fad-aefd-991735717193" />

<img width="1058" height="366" alt="Image" src="https://github.com/user-attachments/assets/16972d57-7bbf-4464-9cc4-d7cac5a8f4c9" />

<img width="1036" height="835" alt="Image" src="https://github.com/user-attachments/assets/44dfc969-37a7-4eac-80fc-ed9f2ccf3ffa" />

If you want to watch the progress, you can use the network view here on github [from ipv6 branch on my fork](https://github.com/asmcc/hestiacp/network). But please be patient. I have also my job, house, family, childern and a lot of other OSS and home automation projects. This IPV6 work here is only one small part of all these things in my life. And that is "one man project" here. Nobody from HESTIA core team works actually on this topic, as far as I know.
